tests/run/typeclass-derivation3.scala

```diff
@@ -0,0 +1,305 @@
+import scala.annotation.tailrec
+
+object datatypes {
+  import typeclasses._
+
+  // An algebraic datatype
+  enum Lst[+T] derives Eq, Pickler, Show {
+    case Cons(hd: T, tl: Lst[T])
+    case Nil
+  }
+  object Lst {}
+
+
+  // A simple product type
+  case class Pair[T](x: T, y: T) derives Eq, Pickler, Show
+
+  // another ADT
+  sealed trait Either[+L, +R] extends Product derives Eq, Pickler, Show
+  case class Left[L](x: L) extends Either[L, Nothing]
+  case class Right[R](x: R) extends Either[Nothing, R]
+
+  //object Either {}
+}
+
+object typeclasses {
+  // A typeclass
+  trait Eq[T] {
+    def eql(x: T, y: T): Boolean
+  }
+
+  object Eq {
+    import scala.typelevel.erasedValue
+    import typelevel._
+
+    inline def tryEql[T](x: T, y: T) = implicit match {
+      case eq: Eq[T] => eq.eql(x, y)
+    }
+
+    inline def eqlElems[Elems <: Tuple](xs: Mirror, ys: Mirror, n: Int): Boolean =
+      inline erasedValue[Elems] match {
+        case _: (elem *: elems1) =>
+          tryEql[elem](xs(n).asInstanceOf, ys(n).asInstanceOf) &&
+          eqlElems[elems1](xs, ys, n + 1)
+        case _: Unit =>
+          true
+      }
+
+    inline def eqlCase[T, Elems <: Tuple](r: Reflected[T], x: T, y: T) =
+      eqlElems[Elems](r.reflect(x), r.reflect(y), 0)
+
+    inline def eqlCases[T, Alts <: Tuple](r: Reflected[T], x: T, y: T): Boolean =
+      inline erasedValue[Alts] match {
+        case _: (Shape.Case[alt, elems] *: alts1) =>
+          x match {
+            case x: `alt` =>
+              y match {
+                case y: `alt` => eqlCase[T, elems](r, x, y)
+                case _ => false
+              }
+            case _ => eqlCases[T, alts1](r, x, y)
+        }
+      case _: Unit =>
+        false
+    }
+
+    inline def derived[T, S <: Shape](implicit ev: Shaped[T, S]): Eq[T] = new {
+      def eql(x: T, y: T): Boolean = inline erasedValue[S] match {
+        case _: Shape.Cases[alts] =>
+          eqlCases[T, alts](ev, x, y)
+        case _: Shape.Case[_, elems] =>
+          eqlCase[T, elems](ev, x, y)
+      }
+    }
+
+    implicit object IntEq extends Eq[Int] {
+      def eql(x: Int, y: Int) = x == y
+    }
+  }
+
+  // Another typeclass
+  trait Pickler[T] {
+    def pickle(buf: mutable.ListBuffer[Int], x: T): Unit
+    def unpickle(buf: mutable.ListBuffer[Int]): T
+  }
+
+  object Pickler {
+    import scala.typelevel.{erasedValue, constValue}
+    import typelevel._
+
+    def nextInt(buf: mutable.ListBuffer[Int]): Int = try buf.head finally buf.trimStart(1)
+
+    inline def tryPickle[T](buf: mutable.ListBuffer[Int], x: T): Unit = implicit match {
+      case pkl: Pickler[T] => pkl.pickle(buf, x)
+    }
+
+    inline def pickleElems[Elems <: Tuple](buf: mutable.ListBuffer[Int], elems: Mirror, n: Int): Unit =
+      inline erasedValue[Elems] match {
+        case _: (elem *: elems1) =>
+          tryPickle[elem](buf, elems(n).asInstanceOf[elem])
+          pickleElems[elems1](buf, elems, n + 1)
```

Comment:
That means we'd have to handle extractors in the inlining framework, which is not yet possible.

Reply:
OK, so maybe forget about having an extractor. Still, I think it would be better to provide the typeclass implementers a correctly typed `elem` value along with its label, without requiring them to manage an elements counter.

Reply:
Yes, I can see that as part of a framework, which could also be built on top of what we have here.